fem/bilininteg_convection.cpp

```diff
@@ -0,0 +1,497 @@
+     }
+   }
+   else
+   {
+     if(dim == 2)
+     {
+       for(int e=0; e<NE; ++e)
+       {
+         ElementTransformation& Te = *fes.GetElementTransformation(e);
+          for(int q=0; q<nq; ++q)
+          {
+            Q->Eval(e_coeff, Te, ir->IntPoint(q));
+            h_C(0,q,e) = alpha*e_coeff(0);
+            h_C(1,q,e) = alpha*e_coeff(1);
+          }
+       }
+     }
+
+     if(dim == 3)
+     {
+       for(int e=0; e<NE; ++e)
+       {
+         ElementTransformation& Te = *fes.GetElementTransformation(e);
+          for(int q=0; q<nq; ++q)
+          {
+            Q->Eval(e_coeff, Te, ir->IntPoint(q));
+            h_C(0,q,e) = alpha*e_coeff(0);
+            h_C(1,q,e) = alpha*e_coeff(1);
+            h_C(2,q,e) = alpha*e_coeff(2);
+          }
+       }
+     }
+
+   }//vector coefficient
+
+
+   auto C = Reshape(coeff.Read(),dim,nq, ne);
+
+   if(dim==2)
+   {
+     auto w = ir->GetWeights().Read();
+     auto J = Reshape(geom->J.Read(), NQ,2,2,NE);
+     auto v = Reshape(pa_data.Write(), 2, NQ, NE);
+
+    MFEM_FORALL(e, NE,
+    {
+      for(int q=0; q<NQ; ++q)
+      {
+        const double J11 = J(q,0,0,e);
+        const double J21 = J(q,1,0,e);
+        const double J12 = J(q,0,1,e);
+        const double J22 = J(q,1,1,e);
+
+        const double w_coeff = w[q];
+
+        double cx = C(0,q,e);
+        double cy = C(1,q,e);
+        v(0,q,e) =    w_coeff*(cx * J22 - cy * J12);
+        v(1,q,e) =  - w_coeff*(cx * J21 - cy * J11);
+      }
+    });
+
+   }//dim = 2
+
+   if(dim==3)
+   {
+     auto w = ir->GetWeights().Read();
+     auto J = Reshape(geom->J.Read(), NQ,3,3,NE);
+     auto v = Reshape(pa_data.Write(), 3, NQ, NE);
+
+    MFEM_FORALL(e, NE,
+    {
+      for(int q=0; q<NQ; ++q) {
+
+        const double J00 = J(q,0,0,e);
+        const double J01 = J(q,0,1,e);
+        const double J02 = J(q,0,2,e);
+
+        const double J10 = J(q,1,0,e);
+        const double J11 = J(q,1,1,e);
+        const double J12 = J(q,1,2,e);
+
+        const double J20 = J(q,2,0,e);
+        const double J21 = J(q,2,1,e);
+        const double J22 = J(q,2,2,e);
+
+        const double A00 = (J11 * J22) - (J12 * J21);
+        const double A01 = (J02 * J21) - (J01 * J22);
+        const double A02 = (J01 * J12) - (J02 * J11);
+
+        const double A10 = (J12 * J20) - (J10 * J22);
+        const double A11 = (J00 * J22) - (J02 * J20);
+        const double A12 = (J02 * J10) - (J00 * J12);
+
+        const double A20 = (J10 * J21) - (J11 * J20);
+        const double A21 = (J01 * J20) - (J00 * J21);
+        const double A22 = (J00 * J11) - (J01 * J10);
+
+        const double w_coeff = w[q];
+
+        double cx = C(0,q,e);
+        double cy = C(1,q,e);
+        double cz = C(2,q,e);
+
+        v(0,q,e) = w_coeff*(cx*A00 + cy*A01 + cz*A02);
+        v(1,q,e) = w_coeff*(cx*A10 + cy*A11 + cz*A12);
+        v(2,q,e) = w_coeff*(cx*A20 + cy*A21 + cz*A22);
+      }
+
+    });
+
+   }//dim = 3
+   mesh->DeleteGeometricFactors();
```

Comment:
The geometric factors should not be deleted by integrators -- we want different integrators to be able to reuse them.

This call should be done only when the mesh is modified in a way that invalidates the geometric factors, e.g. mesh refinement, or updated node positions.

Reply:
👍

Reply:
Is there a way to trigger the recalculation of geometric factors (for problems where the coefficients depend on time)? Unless we delete them I find that we end up with the initial ones.

Reply:
The geometric factors returned by the mesh do not include any coefficients -- just purely geometric information. If a coefficient is changing, just re-assemble the bilinear form.